### PR TITLE
CASSANDRA-20330 check invalid constraint combinations

### PR DESCRIPTION
**Pull request by smiklosovic**
Thanks for sending a pull request! Here are some tips if you're new here:
 
 * Ensure you have added or run the [appropriate tests](https://cassandra.apache.org/_/development/testing.html) for your PR.
 * Be sure to keep the PR description updated to reflect all changes.
 * Write your PR title to summarize what this PR proposes.
 * If possible, provide a concise example to reproduce the issue for a faster review.
 * Read our [contributor guidelines](https://cassandra.apache.org/_/development/index.html)
 * If you're making a documentation change, see our [guide to documentation contribution](https://cassandra.apache.org/_/development/documentation.html)
 
Commit messages should follow the following format:

```
<One sentence description, usually Jira title or CHANGES.txt summary>

<Optional lengthier description (context on patch)>

patch by <Authors>; reviewed by <Reviewers> for CASSANDRA-#####

Co-authored-by: Name1 <email1>
Co-authored-by: Name2 <email2>

```

The [Cassandra Jira](https://issues.apache.org/jira/projects/CASSANDRA/issues/)

### CODE REVIEW

File: src/java/org/apache/cassandra/cql3/constraints/ColumnConstraints.java [hunk 1]
```diff
@@ -101,9 +119,87 @@ public boolean hasRelevantConstraints()
...
+
+            if (columnConstraint.getConstraintType() == ConstraintType.SCALAR)
+                scalarConstraints.add(columnConstraint);
+        }
+
+        if (!duplicateConstraints.isEmpty())
+            throw new InvalidConstraintDefinitionException(format("There are duplicate constraint definitions on column '%s': %s",
+                                                                  columnName,
+                                                                  duplicateConstraints));
+        if (scalarConstraints.size() > 2)
```
Comment (bbotella) [inline src/java/org/apache/cassandra/cql3/constraints/ColumnConstraints.java]: Not so sure about this particular one. Would something like:

`CHECK val > -5 AND val !=0 AND val < 5`? Or `CHECK val != -5 AND val !=0 AND val != 5`? I think there may be cases for more than 2 scalar constraints

File: src/java/org/apache/cassandra/cql3/constraints/ColumnConstraints.java [hunk 2]
```diff
@@ -101,9 +119,87 @@ public boolean hasRelevantConstraints()
...
+            throw new InvalidConstraintDefinitionException(format("There can not be more than 2 scalar constraints on a column '%s' but you have specified %s",
+                                                                  columnName,
+                                                                  scalarConstraints.size()));
+
+        for (ColumnConstraint<?> scalar : scalarConstraints)
+        {
+            Operator operator = ((ScalarColumnConstraint) scalar).relationType();
+
+            if (!ScalarColumnConstraint.SUPPORTED_OPERATORS.contains(operator))
```
Comment (bbotella) [inline src/java/org/apache/cassandra/cql3/constraints/ColumnConstraints.java]: I think this check should be done in the `validate` method of the ScalarColumnConstraint

File: src/java/org/apache/cassandra/cql3/constraints/ColumnConstraints.java [hunk 3]
```diff
@@ -101,9 +119,110 @@ public boolean hasRelevantConstraints()
...
+                if (scalarColumnConstraint.relationType() == NEQ)
+                    notEquals.add(scalarColumnConstraint);
+            }
+        }
+
+        if (!duplicateConstraints.isEmpty())
+            throw new InvalidConstraintDefinitionException(format("There are duplicate constraint definitions on column '%s': %s",
+                                                                  columnName,
+                                                                  duplicateConstraints));
+        if ((scalarConstraints.size() - notEquals.size() > 2))
```
Comment (bbotella) [inline src/java/org/apache/cassandra/cql3/constraints/ColumnConstraints.java]: Nice. That last commit fixes the comment I was writing here :-)

File: src/java/org/apache/cassandra/cql3/constraints/ColumnConstraints.java [hunk 4]
```diff
@@ -101,9 +119,110 @@ public boolean hasRelevantConstraints()
...
+        for (ColumnConstraint<?> columnConstraint : constraints)
+        {
+            String constraintName = columnConstraint.name();
+            if (!constraintNames.add(constraintName))
+            {
+                if (columnConstraint.getConstraintType() == ConstraintType.SCALAR)
+                {
+                    ScalarColumnConstraint scalarColumnConstraint = (ScalarColumnConstraint) columnConstraint;
+                    Operator operator = scalarColumnConstraint.relationType();
+                    if (operator != NEQ)
```
Comment (bbotella) [inline src/java/org/apache/cassandra/cql3/constraints/ColumnConstraints.java]: I think this check shouldn't only be added for scalar constraints. On checking duplicates by name, for function constraints, the relation is ignored, and I think something like this should also be valid:
`CHECK LENGTH(val) != 0 AND LENGTH(val) != 10`

Take this test for example:
```
@Test
    public void testLength() throws Throwable
    {
        FunctionColumnConstraint l1 = new FunctionColumnConstraint.Raw(new ColumnIdentifier("LENGTH", false),
                                                                       new ColumnIdentifier("val", true),
                                                                       NEQ,
                                                                       "10").prepare();

        FunctionColumnConstraint l2 = new FunctionColumnConstraint.Raw(new ColumnIdentifier("LENGTH", false),
                                                                       new ColumnIdentifier("val", true),
                                                                       NEQ,
                                                                       "0").prepare();


        ColumnConstraints constraints = new ColumnConstraints(List.of(l1, l2));
        constraints.checkInvalidConstraintsCombinations(new ColumnIdentifier("val", true), UTF8Type.instance);
    }
```

File: test/unit/org/apache/cassandra/contraints/ScalarSatisfiabilityTest.java [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+import static org.apache.cassandra.cql3.Operator.EQ;
+import static org.apache.cassandra.cql3.Operator.GT;
+import static org.apache.cassandra.cql3.Operator.GTE;
+import static org.apache.cassandra.cql3.Operator.LT;
+import static org.apache.cassandra.cql3.Operator.LTE;
+import static org.apache.cassandra.cql3.Operator.NEQ;
+import static org.apache.cassandra.cql3.constraints.ScalarColumnConstraint.SUPPORTED_OPERATORS;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+public class ScalarSatisfiabilityTest
```
Comment (bbotella) [inline test/unit/org/apache/cassandra/contraints/ScalarSatisfiabilityTest.java]: We also need tests for LENGHT and NOT NULL constraints.

File: test/unit/org/apache/cassandra/contraints/ScalarSatisfiabilityTest.java [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+        new ColumnConstraints(of(get(LT, 5), get(GT, 0))).checkInvalidConstraintsCombinations(columnIdentifier, IntegerType.instance);
+
+        // three - invalid
+
+        assertThatThrownBy(() -> new ColumnConstraints(of(get(LT, 5), get(GT, 0), get(GTE, 0)))
+                                 .checkInvalidConstraintsCombinations(columnIdentifier, IntegerType.instance))
+        .hasMessage("There can not be more than 2 scalar constraints (not including non-equal relations) on a column 'a_column' but you have specified 3");
+
+        // valid
+        new ColumnConstraints(of(get(LT, 5), get(GT, 0), get(NEQ, 3)))
```
Comment (bbotella) [inline test/unit/org/apache/cassandra/contraints/ScalarSatisfiabilityTest.java]: Add a 
```
new ColumnConstraints(of(get(EQ, 5), get(EQ, 0)))
        .checkInvalidConstraintsCombinations(columnIdentifier, IntegerType.instance);
```

case

File: src/java/org/apache/cassandra/cql3/constraints/ColumnConstraints.java (hunk 1 above)
Comment (smiklosovic) [reply src/java/org/apache/cassandra/cql3/constraints/ColumnConstraints.java]: @bbotella That should be possible. This patch enables that already, no? We can have more scalars than two. Why does it show that you are commenting on "outdated" code?

File: src/java/org/apache/cassandra/cql3/constraints/ColumnConstraints.java (hunk 3 above)
Comment (smiklosovic) [reply src/java/org/apache/cassandra/cql3/constraints/ColumnConstraints.java]: yeah.

File: src/java/org/apache/cassandra/cql3/constraints/ColumnConstraints.java (hunk 4 above)
Comment (smiklosovic) [reply src/java/org/apache/cassandra/cql3/constraints/ColumnConstraints.java]: Then we would need to have a register of constraints which support `!=` operator and iterate over them.

I think that similarly to scalar, length would also need to declare what operators it is acting on.

File: test/unit/org/apache/cassandra/contraints/ScalarSatisfiabilityTest.java (hunk 6 above)
Comment (smiklosovic) [reply test/unit/org/apache/cassandra/contraints/ScalarSatisfiabilityTest.java]: @bbotella that is covered already, this is a case of duplicity.

File: src/java/org/apache/cassandra/cql3/constraints/ColumnConstraints.java (hunk 1 above)
Comment (bbotella) [reply src/java/org/apache/cassandra/cql3/constraints/ColumnConstraints.java]: Yes, I posted this comment before your last commit. Disregard

File: src/java/org/apache/cassandra/cql3/constraints/AbstractFunctionSatisfiabilityChecker.java [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+     * @param filteredConstraints pair of all constraints and all constraints having not-equal operator
+     */
+    private void checkNumberOfConstraints(ColumnMetadata columnMetadata, Pair<List<CONSTRAINT_TYPE>, List<CONSTRAINT_TYPE>> filteredConstraints)
+    {
+        List<? extends AbstractFunctionConstraint<CONSTRAINT_TYPE>> allConstraints = filteredConstraints.left;
+        List<? extends AbstractFunctionConstraint<CONSTRAINT_TYPE>> notEqualConstraints = filteredConstraints.right;
+
+        if ((allConstraints.size() - notEqualConstraints.size() > 2))
+        {
+            throw new InvalidConstraintDefinitionException(format("There can not be more than 2 constraints (not including non-equal relations) on a column '%s' but you have specified %s",
```
Comment (bbotella) [inline src/java/org/apache/cassandra/cql3/constraints/AbstractFunctionSatisfiabilityChecker.java:97]: This has me worried after the counter example we discussed offline:
```
alter ks.tb alter val check json(val) and length(val) < 1000 and length(val)>100
```

Should be a valid constraint.

File: src/java/org/apache/cassandra/cql3/constraints/ColumnConstraint.java [hunk 8]
```diff
@@ -46,32 +53,57 @@ public enum ConstraintType
...
+            for (ConstraintType constraintType : ConstraintType.values())
+                result.addAll(Arrays.asList(constraintType.satisfiabilityCheckers));
+
+            return result.toArray(new SatisfiabilityChecker[0]);
+        }
+    }
+
+    public abstract String name();
+
+    public String fullName()
```
Comment (bbotella) [inline src/java/org/apache/cassandra/cql3/constraints/ColumnConstraint.java:98]: We probably don't need this method

File: src/java/org/apache/cassandra/cql3/constraints/ColumnConstraint.java [hunk 9]
```diff
@@ -46,32 +53,57 @@ public enum ConstraintType
...
 
     public abstract void appendCqlTo(CqlBuilder builder);
 
+    public abstract boolean enablesDuplicateDefinitions(String name);
+
     /**
      * Method that evaluates the condition. It can either succeed or throw a {@link ConstraintViolationException}.
      *
-     * @param valueType value type of the column value under test
+     * @param valueType   value type of the column value under test
```
Comment (bbotella) [inline src/java/org/apache/cassandra/cql3/constraints/ColumnConstraint.java:112]: Remove extra space

File: src/java/org/apache/cassandra/cql3/constraints/AbstractFunctionSatisfiabilityChecker.java (hunk 7 above)
Comment (smiklosovic) [reply src/java/org/apache/cassandra/cql3/constraints/AbstractFunctionSatisfiabilityChecker.java:97]: @bbotella that will be all possible. This check is for _one particular constraint function_ which is extending `AbstractFunctionConstraint`.

So it will be treated for e.g. scalar function, then it will be treated for lenght specifically ... for any function, separately.

json function will have its own "satisfiability checker" and for that particular function, it is unary function, and it will be always satisfied. We do not need to look into any other function to decide if it is satisfiable or not.

File: src/java/org/apache/cassandra/cql3/constraints/ColumnConstraint.java (hunk 8 above)
Comment (smiklosovic) [reply src/java/org/apache/cassandra/cql3/constraints/ColumnConstraint.java:98]: @bbotella 

We need to make a difference between "name of the function" which is just `LENGTH`, for example, and "full name" which also contains its operator. That is necessary to realize if there is a duplicate or not.

`a CHECK LENGTH(a) > 0 AND LENGTH(a) > 10`

This means that "full name" will be `LENGTH >` and `LENGTH >`.

That means that we see that we have used same function with same operator twice - so we fail. This is what `DuplicatesChecker` is doing.

If not done in DuplicatesChecker, we would need to check for duplicities in every specific satisfaction checker all the time all over again.

File: src/java/org/apache/cassandra/cql3/constraints/ColumnConstraint.java (hunk 9 above)
Comment (smiklosovic) [reply src/java/org/apache/cassandra/cql3/constraints/ColumnConstraint.java:112]: @bbotella  This is how IDEA formats it, it will align the comments below each other, depending on what the longest parameter name is for that method.

File: src/java/org/apache/cassandra/cql3/constraints/ColumnConstraint.java (hunk 8 above)
Comment (smiklosovic) [reply src/java/org/apache/cassandra/cql3/constraints/ColumnConstraint.java:98]: I have added documentation to that method.

**Comment by bbotella** [comment]
+1 (nb)

**Comment by bbotella** [comment]
+1
